core/trino-main/src/main/java/io/trino/security/AccessControl.java

```diff
@@ -395,63 +395,63 @@ default void checkCanSetViewAuthorization(SecurityContext context, QualifiedObje
      *
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanCreateRole(SecurityContext context, String role, Optional<TrinoPrincipal> grantor, String catalogName);
+    void checkCanCreateRole(SecurityContext context, String role, Optional<TrinoPrincipal> grantor, Optional<String> catalogName);
 
     /**
      * Check if identity is allowed to drop the specified role.
      *
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanDropRole(SecurityContext context, String role, String catalogName);
+    void checkCanDropRole(SecurityContext context, String role, Optional<String> catalogName);
 
     /**
      * Check if identity is allowed to grant the specified roles to the specified principals.
      *
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanGrantRoles(SecurityContext context, Set<String> roles, Set<TrinoPrincipal> grantees, boolean adminOption, Optional<TrinoPrincipal> grantor, String catalogName);
+    void checkCanGrantRoles(SecurityContext context, Set<String> roles, Set<TrinoPrincipal> grantees, boolean adminOption, Optional<TrinoPrincipal> grantor, Optional<String> catalogName);
 
     /**
      * Check if identity is allowed to revoke the specified roles from the specified principals.
      *
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanRevokeRoles(SecurityContext context, Set<String> roles, Set<TrinoPrincipal> grantees, boolean adminOption, Optional<TrinoPrincipal> grantor, String catalogName);
+    void checkCanRevokeRoles(SecurityContext context, Set<String> roles, Set<TrinoPrincipal> grantees, boolean adminOption, Optional<TrinoPrincipal> grantor, Optional<String> catalogName);
 
     /**
      * Check if identity is allowed to set role for specified catalog.
      *
      * @throws AccessDeniedException if not allowed
      */
-    void checkCanSetRole(SecurityContext context, String role, String catalogName);
+    void checkCanSetCatalogRole(SecurityContext context, String role, String catalogName);
```

Comment:
For system roles, we don't do a check like this, because it would require the security system to know which role grants the user has.  Since role grants are kept in the metadata system this requires that there be a coupling between metadata and security.  Instead, when setting a role, we simply get the granted system roles from the metadata, and then we check if that list contains the specified role.  I expect this will be ok since the default role setting is "all roles", so the get roles method must be fast (so no perf concerns), and this keeps metadata separate from the security system.

Reply:
> Notice that the problem already exists on catalog level.
Yep, but catalog roles have tons of problems, but I work hard in this PR to not touch catalog roles.
> Notice that for catalogs we could simply get the granted catalog roles from the metadata, and then we check if that list contains the specified role
Someone wanted to do that in another PR, but at that point they should take a long look at catalog roles overall.